### ISSUE TITLE
ExpandoObject property doesn't exist

### ISSUE THREAD

**Issue by elvogel** [microsoft/RulesEngine#42]
I'm trying something different based on something you mentioned in the last issue I posted, so I thought I'd post this in a separate thread. 

I need to do some OData lookups; as a result, my `Utils` object is now non-static and has a constructor so I can cache some things, etc.. This object is called `utils` and my ExpandoObject is `input1`. 

When I run the rules engine, my code looks like this:

```csharp
foreach (var entity in entities)
{
    var serializer = new DynamicTableEntityJsonSerializer();
    var converter = new ExpandoObjectConverter();

    //serialize the entity as JSON, then deserialize as an ExpandoObject
    var data = serializer.Serialize(entity);
    dynamic input1 = JsonConvert.DeserializeObject<ExpandoObject>(data, converter);

    foreach (var rule in workflowRules)
    {
        List<RuleResultTree> results;
        try
        {
            results = engine.ExecuteRule(rule.WorkflowName, new RuleParameter("utils",utils),new RuleParameter("input1",input1));
        }
        catch (Exception e)
        {
            log.LogError($"[{tableName}] {e.GetType()}: {e.Message}\r\n{e.StackTrace}");
            throw new ApplicationException($"An exception was thrown in the rules engine at row {row} for {tableName}. Check the rules and try again.");
        }

        foreach (var result in results)
        {
            if (result.IsSuccess) continue;

            var sb = new StringBuilder();
            sb.AppendLine($"{rule.WorkflowName}.{result.Rule} encountered an error in row {row}: {result.ExceptionMessage}");
            var msg = result.GetMessages();
            foreach (var warn in msg.WarningMessages)
            {
                sb.AppendLine($"Warning: {warn}");
            }

            foreach (var errorMessage in msg.ErrorMessages)
            {
                sb.AppendLine($"Error: {errorMessage}");
            }
            log.LogError(sb.ToString());
            log.LogError($"processing for {tableName} will now stop due to errors.");
            throw new ApplicationException($"Processing for {tableName} stopped because rule {result.Rule.RuleName} failed to pass.");
        }
    }

    row++;
}
```

Currently, I'm getting an error that says `ConsItemPricePurchEntity.RulesEngine.Models.Rule encountered an error in row 1: No property or field 'ELEMENTNAME' exists in type 'ExpandoObject'. When I look, the `ELEMENTNAME` field is there with a non-null value. 

So two questions:
- is it allowable to pass in a utility object in this manner? 
- is this exception part of an error? 

Thanks!

**Comment by elvogel** [comment]
The rules:
```json

[
  {
    "WorkflowName": "ConsItemPricePurchEntity",
    "Rules": [
      {
        "RuleName": "VendorPartyNumberExists",
        "SuccessEvent": "10",
        "WarningMessage": "Vendor party number does not exist in Dynamics",
        "ErrorType": "Warning",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "utils.ElementNameExists(input1.ELEMENTNAME) == true"
      },
      {
        "RuleName": "BuildingTypeIdExists",
        "SuccessEvent": "15",
        "WarningMessage": "BuildingTypeId doesn't exist in Dynamics",
        "ErrorType": "Warning",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "utils.BuildingTypeIdExists(input1.BUILDINGTYPEID) == true"
      }
    ]
  }
]

```

**Comment by abbasc52** [comment]
> So two questions:
> 
>     * is it allowable to pass in a utility object in this manner?

Yes it is fine to pass utility this way

> 
>     * is this exception part of an error?

The exception was fixed as a part of v2.1.0 - https://github.com/microsoft/RulesEngine/pull/33 
For version 2.0.0, you can fix it by calling our utility function while passing RuleParameter
```
var typedInput1 = RulesEngine.HelperFunctions.Utils.GetTypedObject(input1);
results = engine.ExecuteRule(rule.WorkflowName, new RuleParameter("utils",utils),new RuleParameter("input1",typedInput1));

```

**Comment by elvogel** [comment]
When using version 2.1.0 of the library, I execute this:
```csharp
results = engine.ExecuteRule(rule.WorkflowName, new RuleParameter("utils",utils),new RuleParameter("input1",input1));
```
I get this:

```csharp
System.ArgumentNullException: Value cannot be null. (Parameter 'input')
   at System.Text.RegularExpressions.Regex.Matches(String input)
   at System.Text.RegularExpressions.Regex.Matches(String input, String pattern, RegexOptions options, TimeSpan matchTimeout)
   at System.Text.RegularExpressions.Regex.Matches(String input, String pattern)
   at RulesEngine.RulesEngine.FormatErrorMessages(IEnumerable`1 result)
   at RulesEngine.RulesEngine.ExecuteRuleByWorkflow(String workflowName, RuleParameter[] ruleParameters)
   at RulesEngine.RulesEngine.ValidateWorkflowAndExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at SM.Integration.Repos.BRE.RulesProcessor.ProcessRules() in C:\dev\sm\develop-lucas\dmvalidator-d365\SM.Integration.Repos\BRE\RulesProcessor.cs:line 92
```
I also tried running it without the `RuleParameter` wrapper for the arguments, and I get the same error.

If I roll back to version 2.0.0, I get the `ExpandoObject` error when I use the `RuleParameter` arguments; when I take them off and pass in the parameters 'raw', it complains that `utils` does not have a property called `ELEMENTNAME`.

**Comment by joshidp** [comment]
@elvogel Can you please explain why WarningMessage is being used in your Rule instead of ErrorMessage property?

**Comment by elvogel** [comment]
I was toying with having the rule emitting a warning instead of an error.

**Comment by abbasc52** [comment]
This issue is resolved in v2.1.1
